lib/zoho_resource_client.rb

```diff
@@ -0,0 +1,73 @@
+class ZohoResourceClient
+  ZOHO_ORG_ID = ArchiveConfig.ZOHO_ORG_ID
+  CONTACT_SEARCH_ENDPOINT = 'https://desk.zoho.com/api/v1/contacts/search'.freeze
+  CONTACT_CREATE_ENDPOINT = 'https://desk.zoho.com/api/v1/contacts'.freeze
+  TICKET_CREATE_ENDPOINT = 'https://desk.zoho.com/api/v1/tickets'.freeze
+
+  def initialize(access_token:, email:, username: nil)
+    @access_token = access_token
+    @email = email
+    @username = username
+  end
+
+  def retrieve_contact_id
+    (find_contact || create_contact).fetch('id')
+  end
+
+  def create_ticket(ticket_attributes:)
+    response_raw = HTTParty.post(
+      TICKET_CREATE_ENDPOINT,
+      headers: headers,
+      body: ticket_attributes.to_json
+    )
+    JSON.parse(response_raw.body)
+  end
+
+  private
+
+  def find_contact
+    response_raw = HTTParty.get(
+      CONTACT_SEARCH_ENDPOINT + '?' + query_string(search_params),
+      headers: headers,
+    )
+    return if response_raw.nil?
+
+    JSON.parse(response_raw.body).fetch('data').first
+  end
+
+  def create_contact
+    response_raw = HTTParty.post(
+      CONTACT_CREATE_ENDPOINT,
+      headers: headers,
+      body: contact_body.to_json
+    )
+    JSON.parse(response_raw.body)
+  end
+
+  def search_params
+    {
+      email: @email,
+      limit: 1,
+      sortBy: 'modifiedTime',
+    }
+  end
+
+  def headers
+    {
+      "Content-Type" => "application/json",
+      "orgId" => ZOHO_ORG_ID,
+      "Authorization" => "Zoho-oauthtoken #{@access_token}"
+    }
+  end
+
+  def contact_body
+    {
+      "lastName" => @username.presence || @email,
+      "email" => @email,
+    }
+  end
+
+  def query_string(params)
+    params.map {|k, v| "#{k}=#{v}" }.join('&')
```

Comment:
Do these query parameters need to be encoded, since they include the user-provided email (which can include `=`, `+`, and `&`)?

Also, emails can contain `*`, which is the character that Zoho uses for wildcard searches. So that might need to be escaped separately in order to ensure that you retrieve the right contact with your search. (Assuming Zoho lets you escape wildcards. It stands to reason that they ought to, though I couldn't find a mention of it when I skimmed the search docs.)

(Side note: If Zoho does let you escape `*` using `\*`, it's worth noting that you can use the `\` character in email addresses if it's somewhere inside a double-quoted section. So you might or might not need to escape that, too.)